mkosi/run.py

```diff
@@ -0,0 +1,313 @@
+
+    for line in path.read_text().splitlines():
+        name, start, count = line.split(":")
+
+        if name == uid or name == user:
+            break
+    else:
+        die(f"No mapping found for {user or uid} in {path}")
+
+    if int(count) != SUBRANGE:
+        die(f"subuid/subgid range length must be {SUBRANGE}, got {count} for {user or uid} from line '{line}'")
+
+    return int(start)
+
+
+def become_root() -> tuple[int, int]:
+    subuid = read_subrange(Path("/etc/subuid"))
+    subgid = read_subrange(Path("/etc/subgid"))
+
+    event = multiprocessing.Event()
+    pid = os.getpid()
+
+    child = os.fork()
+    if child == 0:
+        event.wait()
+
+        # We map the private UID range configured in /etc/subuid and /etc/subgid into the container using
+        # newuidmap and newgidmap. On top of that, we also make sure to map in the user running mkosi so that
+        # we can run still chown stuff to that user or run stuff as that user which will make sure any
+        # generated files are owned by that user. We don't map to the last user in the range as the last user
+        # is sometimes used in tests as a default value and mapping to that user might break those tests.
+        newuidmap = [
+            "newuidmap", pid,
+            0, subuid, SUBRANGE - 100,
+            SUBRANGE - 100, os.getuid(), 1,
+            SUBRANGE - 100 + 1, subuid + SUBRANGE - 100 + 1, 99
+        ]
+        run((str(x) for x in newuidmap))
+
+        newgidmap = [
+            "newgidmap", pid,
+            0, subgid, SUBRANGE - 100,
+            SUBRANGE - 100, os.getgid(), 1,
+            SUBRANGE - 100 + 1, subgid + SUBRANGE - 100 + 1, 99
+        ]
+        run(str(x) for x in newgidmap)
+
+        sys.stdout.flush()
+        sys.stderr.flush()
+
+        os._exit(0)
+
+    unshare(CLONE_NEWUSER)
+    event.set()
+    os.waitpid(child, 0)
+
+    # By default, we're root in the user namespace because if we were our current user by default, we
+    # wouldn't be able to chown stuff to be owned by root while the reverse is possible.
+    os.setresuid(0, 0, 0)
+    os.setresgid(0, 0, 0)
+    os.setgroups([0])
+
+    return SUBRANGE - 100, SUBRANGE - 100
+
+
+def init_namespace() -> None:
+    unshare(CLONE_NEWNS)
+    run(["mount", "--make-rslave", "/"])
+
+
+def foreground() -> None:
+    # If we're connected to a terminal, put the the child process in a new process group and make that
+    # the foreground process group so that only the child process receives SIGINT.
+    if sys.stdin.isatty():
+        os.setpgrp()
+        old = signal.signal(signal.SIGTTOU, signal.SIG_IGN)
+        os.tcsetpgrp(0, os.getpgrp())
+        signal.signal(signal.SIGTTOU, old)
+
+
+# Stores the exception from a subprocess along with its traceback. We have to do this explicitly because the
+# original traceback object cannot be pickled. When stringified, produces the subprocess stacktrace plus the
+# exception message.
+class RemoteException(Exception):
```

Comment:
```suggestion
class RemoteException(Exception):
    """
    Stores the exception from a subprocess along with its traceback. We have to do this explicitly because the
    original traceback object cannot be pickled. When stringified, produces the subprocess stacktrace plus the
    exception message.
    """
```